crates/uv-client/src/registry_client.rs

```diff
@@ -516,6 +517,68 @@ impl RegistryClient {
                 .into_async_read(),
         ))
     }
+
+    /// Copy authentication from one URL to another URL if applicable.
+    ///
+    /// See [`RegistryClient::should_retain_auth`] for details on when authentication is retained.
+    #[must_use]
+    fn safe_copy_auth(request_url: &Url, mut response_url: Url) -> Url {
+        if RegistryClient::should_retain_auth(request_url, &response_url) {
+            response_url
+                .set_username(request_url.username())
+                .unwrap_or_else(|_| {
+                    warn!("Failed to transfer username to response URL: {response_url}")
+                });
+            response_url
+                .set_password(request_url.password())
+                .unwrap_or_else(|_| {
+                    warn!("Failed to transfer password to response URL: {response_url}")
+                });
+        }
+        response_url
+    }
+
+    /// Determine if authentication information should be retained on a new URL.
+    /// Implements the specification defined in RFC 7235 and 7230.
+    ///
+    /// <https://datatracker.ietf.org/doc/html/rfc7235#section-2.2>
+    /// <https://datatracker.ietf.org/doc/html/rfc7230#section-5.5>
+    fn should_retain_auth(request_url: &Url, response_url: &Url) -> bool {
+        // The "scheme" and "authority" components must match to retain authentication
+
+        // Check the scheme.
+        // Note some clients such as Python's `requests` library allow an upgrade
+        // from `http` to `https` but this is not spec-compliant.
+        // <https://github.com/pypa/pip/blob/75f54cae9271179b8cc80435f92336c97e349f9d/src/pip/_vendor/requests/sessions.py#L133-L136>
+        if request_url.scheme() != response_url.scheme() {
+            return false;
+        }
+
+        // Check the authority, which is composed of the host and port.
+        if request_url.host() != response_url.host() {
+            return false;
+        }
+
+        let default_port = match request_url.scheme() {
+            "http" => Some(80),
+            "https" => Some(443),
+            _ => None,
+        };
+
+        // Check the port
+        if request_url.port() != response_url.port() {
+            if (request_url.port() == default_port || request_url.port().is_none())
+                && (response_url.port() == default_port || response_url.port().is_none())
```

Comment:
Not sure if that's the same

```suggestion
            if !request_url.port().is_some_and(|port| != default_port)
                && !response_url.port().is_some_and(|port| != default_port) 
```

What case does this cover: Is it mainly to cover the case where the `request` has the port `None` but the response comes with e.g. port `80`?

Reply:
This is slightly different... I can try to make it better again. But yeah if you look at the test cases and the specification we are only supposed to check the port if it differs from the default port.